### ISSUE TITLE
cscan fails to parse GL/glext.h

### ISSUE THREAD

**Issue by basiliscos** [jddurand/MarpaX-Languages-C-AST#11]
Hi,

While it parses OpenGL headers (```GL/gl.h```) rather well, but the OpenGL extension headers(```GL/glext.h```) it is incapable to handle:

```
cscan --xml /usr/include/GL/glext.h
2015/01/19 11:51:05 FATAL   5647 Error in SLIF parse: No lexeme found at line 8, column 54
* String before error: edef void ( * PFNGLDRAWRANGEELEMENTSPROC) (GLenum\s
* The error was at line 8, column 54, and at character 0x006d 'm', ...
* here: mode, GLuint start, GLuint end, GLsizei count, GLe
Marpa::R2 exception at /home/basiliscos/perl5/perlbrew/perls/perl-5.20.1/lib/site_perl/5.20.1/MarpaX/Languages/C/AST/Impl.pm line 73.

Last position:

At line 8, column 47

typedef void ( * PFNGLDRAWRANGEELEMENTSPROC) (GLenum mode, GLuint start, GLuint end, GLsizei count, GLenum type, const void *indices);
----------------------------------------------^
Uncaught exception from user code:
	 at /home/basiliscos/perl5/perlbrew/perls/perl-5.20.1/lib/site_perl/5.20.1/MarpaX/Languages/C/AST.pm line 184.
	MarpaX::Languages::C::AST::Util::logCroak("%s\x{a}Last position:\x{a}\x{a}%s%s", "Error in SLIF parse: No lexeme found at line 8, column 54\x{a}* S"..., "At line 8, column 47\x{a}\x{a}typedef void ( * PFNGLDRAWRANGEELEMENTS"..., "") called at /home/basiliscos/perl5/perlbrew/perls/perl-5.20.1/lib/site_perl/5.20.1/MarpaX/Languages/C/AST.pm line 184
	MarpaX::Languages::C::AST::parse(MarpaX::Languages::C::AST=HASH(0x70e0a78), SCALAR(0x3a95d48)) called at /home/basiliscos/perl5/perlbrew/perls/perl-5.20.1/lib/site_perl/5.20.1/MarpaX/Languages/C/AST/Grammar/ISO_ANSI_C_2011/Scan.pm line 478
	MarpaX::Languages::C::AST::Grammar::ISO_ANSI_C_2011::Scan::_ast(MarpaX::Languages::C::AST::Grammar::ISO_ANSI_C_2011::Scan=HASH(0x4571670)) called at /home/basiliscos/perl5/perlbrew/perls/perl-5.20.1/lib/site_perl/5.20.1/MarpaX/Languages/C/AST/Grammar/ISO_ANSI_C_2011/Scan.pm line 123
	MarpaX::Languages::C::AST::Grammar::ISO_ANSI_C_2011::Scan::new("MarpaX::Languages::C::AST::Grammar::ISO_ANSI_C_2011::Scan", "filename_filter", qr(glext\.h), "filename", "/usr/include/GL/glext.h", "asDOM", 1, "enumType", "int", ...) called at /home/basiliscos/perl5/perlbrew/perls/perl-5.20.1/lib/site_perl/5.20.1/MarpaX/Languages/C/Scan.pm line 23
	MarpaX::Languages::C::Scan::new("MarpaX::Languages::C::Scan", "filename", "/usr/include/GL/glext.h", "filename_filter", qr(glext\.h), "asDOM", 1, "enumType", "int", ...) called at /home/basiliscos/perl5/perlbrew/perls/perl-5.20.1/bin/cscan line 146
```

Thanks!
wbr, basiliscos.

**Comment by jddurand** [comment]
Hello,
This file cannot be used stand-alone, it is unparsable without other #includes around it. 
It tried gcc on it directly and it fails as expected.

Nevertheless: 
c2ast have a --lazy option that usually help to parse "incomplete" files. It works in the case of glext.h. But csan does not have such option.

If you want it... tell me, I'll try to add that (cscan is using c2ast internally).

**Comment by basiliscos** [comment]
Hi, 

Yes, I probably will need that feature, but let us see the ```c2ast``` output first,

For now it keeps silence. 

```
c2ast --lazy /usr/include/GL/glext.h
```

And when I invoke it for help, it continues keeping silence

```
c2ast --help
```

**Comment by jddurand** [comment]
``cast --help`` silent ?

With no instruction on output, it will do the ast and nothing else. c2ast is older than cscan and can do only Data::Dumper(), which usually hog the memory as soon as the ast is big, though;

Lazy would not be enough and one would have to instruct c2ast about GL typedefs, e.g.:

``c2ast --progress --typedef GLenum,GLboolean,GLbitfield,GLvoid,GLbyte,GLshort,GLint,GLubyte,GLushort,GLuint,GLsizei,GLfloat,GLclampf,GLdouble,GLclampd glext.h``

Anyway, you are using cscan, and this is good. So I'll try to add some c2ast options into the cscan command-line.

**Comment by basiliscos** [comment]
Yes, please, be so kind to dump XML-ast for glext.h for cscan, i.e. via 

```
cscan --xml --lazy /usr/include/GL/glext.h
```

I expect tha the output for gl.h would be the same with ```--lazy``` option or without it i.e. 

```
cscan --xml --lazy /usr/include/GL/gl.h
```

```
cscan --xml - /usr/include/GL/gl.h
```

PS. ```--progress``` option is cool! I plan to use ```cscan``` for code analysis and further code generation, so generally your tool seems to be very nice for that :)

**Comment by jddurand** [comment]
Fixed with b3fe7115903952b955607fafeee812000e855ce0 and uploaded as MarpaX-Languages-C-AST-0.43-TRIAL that you may want to validate. I have tested with:

```perl -Ilib bin/cscan --trace --xml --typedef GLenum,GLboolean,GLbitfield,GLvoid,GLbyte,GLshort,GLint,GLubyte,GLushort,GLuint,GLsizei,GLfloat,GLclampf,GLdouble,GLclampd /tmp/glext.h```

which takes some time but well.

--lazy option is supported but not needed in this case.

**Comment by basiliscos** [comment]
Thanks! Looks as what I exactly expected what! :+1:
